File: Java/Backend/GameCore/CapivaraWarsCore/src/br/com/capivarawars/core/game/gameobject/Canoa.java
package br.com.capivarawars.core.game.gameobject;

//<editor-fold defaultstate="collapsed" desc="imports...">
import br.com.capivarawars.core.*;
import br.com.capivarawars.core.game.component.*;
import br.com.capivarawars.core.primitive.*;
import br.com.capivarawars.core.primitive.patterns.*;
import java.util.List;
import java.util.ArrayList;
//</editor-fold>

//<editor-fold defaultstate="collapsed" desc="documentation...">
/**
 * Objective: ...
 * 
 * Description: ...
 * 
 * @version 1.0.0
 * @author TPEDROSO, 27/09/2019, 12:22:57
 * Last update: -
 *///</editor-fold>
public class Canoa extends GameObject{
    
    //<editor-fold defaultstate="collapsed" desc="attributes...">
    
    //<editor-fold defaultstate="collapsed" desc="main attributes...">
    private String nome;
    private CorPadrao cor;
    private Pedacos pedacos;    
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="constants attributes...">
    
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="static attributes...">
    
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="auxiliary attributes...">
   
    //</editor-fold>
    
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="constructors...">
    public Canoa(String nome, CorPadrao cor, int tamanho){
        super(Canoa.class);
        this.nome = nome;
        
        pedacos = new Pedacos(tamanho); 
                
        this.cor = cor;
    }
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="methods...">
    
    //<editor-fold defaultstate="collapsed" desc="getter and setter methods...">
    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }    
    
    public CorPadrao getCor(){
        return cor;
    }
    
    public void setCor(CorPadrao novaCor){
        cor = novaCor;
        
        for (int cont = 0; cont < lengthOfPedacos(); cont++) {
            getPedaco(cont).setCor(novaCor);
        }
    }
    
    public Pedaco getPedaco(int indicePedaco){
        if (indicePedaco >= 0 && indicePedaco < lengthOfPedacos()) {
            return pedacos.get(indicePedaco);
        }
        
        return null;
    }
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="override methods...">
    @Override
    public String toString() {
        StringBuilder finalText = new StringBuilder();
        
        finalText
                .append(Canoa.class.getSimpleName())
                .append(' ')
                .append(ImprovableToString.CLASS_OPENING_CHAR)
                .append(toStringWithAttibutesOnly(ImprovableToString.TAB_SIZE, true))
                .append('\n')
                .append(ImprovableToString.CLASS_CLOSING_CHAR);
        
        return finalText.toString();
    }
    
    @Override
    public String toStringWithAttibutesOnly(int tabSizeForEachAttribute, boolean includeParentAttributes) {
        StringBuilder finalText = new StringBuilder(200);
        
        if (includeParentAttributes) {
            finalText.append(super.toStringWithAttibutesOnly_GameObjectDemo(tabSizeForEachAttribute));
        }
        
        
        StringBuilder tabSpace = new StringBuilder();        
        for (int count = 0; count < tabSizeForEachAttribute; count++) {
            tabSpace.append(' ');
        }
        
        finalText
                .append(ImprovableToString.ATTRIBUTE_SEPARATOR)
                .append('\n')
                .append(tabSpace)
                .append("nome: ")
                .append(getNome())
                
                .append(ImprovableToString.ATTRIBUTE_SEPARATOR)
                .append('\n')
                .append(tabSpace)
                .append("cor: ")
                .append(getCor())
                
                .append(ImprovableToString.ATTRIBUTE_SEPARATOR)
                .append('\n')
                .append(tabSpace)
                .append("Pedacos(")
                .append(lengthOfPedacos())
                .append("): ");
        
        for (int count = 0; count < lengthOfPedacos(); count++) {
            if (getPedaco(count).isDestruido()) {
                finalText.append("[~]");
            }
            else{
                finalText.append("[O]");
            }
        }
        
        finalText
                .append(ImprovableToString.ATTRIBUTE_SEPARATOR)
                .append('\n')
                .append(tabSpace)
                .append("isDestruida: ")
                .append(isDestruida());
        
        return finalText.toString();
    }
    //</editor-fold>    
    
    //<editor-fold defaultstate="collapsed" desc="auxiliary methods...">
    
    
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="static methods...">
    
    //</editor-fold>
    
    //<editor-fold defaultstate="collapsed" desc="main methods...">
    public int lengthOfPedacos(){
        return pedacos.length();
    }
    
    public int lengthOfPedacosInteiros(){
        int contPedacosInteiros = 0;
        for (int cont = 0; cont < lengthOfPedacos(); cont++) {
            if (!getPedaco(cont).isDestruido()) {
                contPedacosInteiros++;
            }
        }
        
        return contPedacosInteiros;
    }
    
    public boolean isDestruida(){
        boolean destruida = true;
        
        for (int cont = 0; cont < lengthOfPedacos(); cont++) {
            destruida &= getPedaco(cont).isDestruido();
        }
        
        return destruida;
    }
    
    public boolean destruirPedaco(int indicePedaco){
        if (indicePedaco >= 0 && indicePedaco < lengthOfPedacos()
                && !getPedaco(indicePedaco).isDestruido()) {
            getPedaco(indicePedaco).destruir();
            Pedaco pedacoTemp = pedacos.get(indicePedaco);
            pedacos.remove(indicePedaco);
            pedacos.add(pedacoTemp);
            return true;
        }
        
        return false;
    }
    
    public void destruirTodosPedacos(){
        for (int cont = 0; cont < lengthOfPedacos(); cont++) {
            pedacos.get(cont).destruir();
        }
    }
    
    public void reconstruirTodosPedacos(){
        for (int cont = 0; cont < lengthOfPedacos(); cont++) {
            pedacos.get(cont).construir();
        }
    }
    
    public boolean reconstruirPedaco(int indicePedaco){
        if (indicePedaco >= 0 && indicePedaco < lengthOfPedacos()
                && pedacos.get(indicePedaco).isDestruido()) {
            pedacos.get(indicePedaco).construir();
            Pedaco pedacoTemp = pedacos.get(indicePedaco);
            pedacos.remove(indicePedaco);
            pedacos.addAt(0, pedacoTemp);
            return true;
        }
        
        return false;
    }
    //</editor-fold>
    
    //</editor-fold>
    
}//class

File: Node/capivara-wars-react/src/gamecore/Canoa.js
import Validator from '../tool/Validator';

/**
 * Objective: ...
 * 
 * Description: ...
 * 
 * @version 1.0.0
 * @author tiago, 28/11/2019, 19:48:17
 * Last update: -
 */

export default class Canoa {

	tamanhoMax;
	tamanhoAtual;
	destruida;
	pedacos;

	constructor(tamanho) {
		if (Validator.isInteger(tamanho)) {
			this.tamanhoMax = tamanho;
			this.reiniciar();

			//atualiza atributo da classe
			this.isDestruida();
		}

		//Construtor sem parâmetros
		if (Validator.isUndefined(tamanho)) {
			this.tamanhoMax = 4;
			this.reiniciar();

			//atualiza atributo da classe
			this.isDestruida();
		}
	}

	reiniciar() {
		this.pedacos = new Array(this.getTamanhoMax());

		for (var cont = 0; cont < this.getTamanhoMax(); cont++) {
			this.pedacos[cont] = true;
		}

		this.tamanhoAtual = this.pedacos.length;

		this.destruida = false;
	}

	getTamanhoMax() {
		return this.tamanhoMax;
	}

	getTamanhoAtual() {
		this.tamanhoAtual = this.pedacos.length;
		return this.tamanhoAtual;
	}

	isDestruida() {
		if (this.getTamanhoAtual() === 0) {
			this.destruida = true;
		}
		else {
			this.destruida = false;
		}

		return this.destruida;
	}

	getPedaco(indicePedaco) {
		if (Validator.isInteger(indicePedaco)
			&& indicePedaco >= 0
			&& indicePedaco < this.getTamanhoAtual()) {

			return this.pedacos[indicePedaco];
		}

		return null;
	}

	destruirPedaco() {
		if (!this.isDestruida()) {
			this.pedacos.pop();

			//atualiza atributo da classe
			this.isDestruida();

			return true;
		}

		//atualiza atributo da classe
		this.isDestruida();
		return false;
	}

	destruirTodosPedacos() {
		this.pedacos = [];

		//atualiza atributo da classe
		this.isDestruida();

		this.destruida = true;
	}

	reconstruirTodosPedacos() {
		this.reiniciar();

		//atualiza atributo da classe
		this.isDestruida();

		this.destruida = false;
	}

	reconstruirPedaco() {
		if (this.getTamanhoAtual() < this.getTamanhoMax()) {
			this.pedacos.push(true);
		}

		//atualiza atributo da classe
		this.isDestruida();
	}

}